File: internal/outreach/send.go
package outreach

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/hanifcarroll/linkedin-network-run/internal/app"
)

type SendMessageOptions struct {
	LeadID     string
	RunID      string
	Session    string
	Playwriter string
	Script     string
	OutDir     string
	DryRun     bool
	AllowSend  bool
	TimeoutMS  uint32
}

type MessageCandidate struct {
	ID         string  `json:"id"`
	Name       string  `json:"name"`
	ProfileURL string  `json:"profileUrl"`
	SearchURL  *string `json:"searchUrl,omitempty"`
	Source     string  `json:"source"`
	LeadType   string  `json:"leadType"`
	Title      *string `json:"title,omitempty"`
	Company    *string `json:"company,omitempty"`
}

type MessageSendResult struct {
	Candidate           MessageCandidate `json:"candidate"`
	DryRun              bool             `json:"dryRun"`
	URL                 *string          `json:"url"`
	MessageLength       int              `json:"messageLength"`
	Status              string           `json:"status"`
	Reason              *string          `json:"reason"`
	Action              json.RawMessage  `json:"action"`
	SearchRowAction     json.RawMessage  `json:"searchRowAction"`
	ConversationCheck   json.RawMessage  `json:"conversationCheck"`
	SubjectFill         json.RawMessage  `json:"subjectFill"`
	BodyFill            json.RawMessage  `json:"bodyFill"`
	Send                json.RawMessage  `json:"send"`
	SendButtons         json.RawMessage  `json:"sendButtons"`
	ProfileAPIResponses json.RawMessage  `json:"profileApiResponses"`
	ComposerSelector    *string          `json:"composerSelector"`
	Body                *string          `json:"body"`
}

func SendMessage(store *Store, options SendMessageOptions) error {
	options.RunID = normalizeRunID(options.RunID, "message")
	if options.Session == "" {
		return fmt.Errorf("--session is required")
	}
	resolvedSession, err := resolvePlaywriterSession(options.Playwriter, options.Session)
	if err != nil {
		return err
	}
	options.Session = resolvedSession
	if options.LeadID == "" {
		return fmt.Errorf("--lead-id is required")
	}
	if options.Script == "" {
		return fmt.Errorf("--script is required")
	}
	if options.OutDir == "" {
		return fmt.Errorf("--out-dir is required")
	}
	dryRun := options.DryRun || !options.AllowSend

	state, err := store.Load()
	if err != nil {
		return err
	}
	index := findLeadByID(state.Leads, options.LeadID)
	if index < 0 {
		return fmt.Errorf("unknown lead id %q", options.LeadID)
	}
	lead := state.Leads[index]
	if lead.Draft == nil || cleanText(lead.Draft.Body) == "" {
		return fmt.Errorf("lead %s has no draft; run draft first", lead.ID)
	}
	if lead.ProfileURL == nil || cleanText(*lead.ProfileURL) == "" {
		return fmt.Errorf("lead %s has no profile URL", lead.ID)
	}
	if !dryRun && lead.MessageStatus != MessageStatusDryRunReady {
		return fmt.Errorf("lead %s is %s; real sends require %s", lead.ID, lead.MessageStatus, MessageStatusDryRunReady)
	}
	if !dryRun && !leadMatchesSendableBucket(state, lead, bucketForLead(lead)) {
		return fmt.Errorf("lead %s is not sendable for %s; agency sends require qualified account context", lead.ID, bucketForLead(lead))
	}
	if !dryRun && !options.AllowSend {
		return fmt.Errorf("real send requires --allow-send")
	}
	if err := os.MkdirAll(options.OutDir, 0o755); err != nil {
		return fmt.Errorf("creating %s: %w", options.OutDir, err)
	}
	outPath := filepath.Join(options.OutDir, lead.ID+".json")
	candidate := MessageCandidate{
		ID:         lead.ID,
		Name:       lead.Name,
		ProfileURL: *lead.ProfileURL,
		Source:     lead.Source,
		LeadType:   string(lead.LeadType),
		Title:      lead.Title,
		Company:    lead.Company,
	}
	if cursor, ok := state.CaptureCursors[lead.Source]; ok && cursor.ResumeURL != nil && cleanText(*cursor.ResumeURL) != "" {
		candidate.SearchURL = cursor.ResumeURL
	}
	config := map[string]any{
		"candidate": candidate,
		"message":   lead.Draft.Body,
		"subject":   draftSubject(lead),
		"out":       outPath,
		"dryRun":    dryRun,
		"allowSend": options.AllowSend,
	}
	rawConfig, err := json.Marshal(config)
	if err != nil {
		return err
	}
	configJS := fmt.Sprintf("state.recruiterAgencyMessageConfig = %s; console.log(JSON.stringify(state.recruiterAgencyMessageConfig));", string(rawConfig))
	if err := app.RunPlaywriterConfig(options.Playwriter, options.Session, configJS); err != nil {
		return err
	}
	if err := app.RunPlaywriterFileWithTimeout(options.Playwriter, options.Session, options.Script, options.TimeoutMS); err != nil {
		return err
	}
	result, err := LoadMessageSendResult(outPath)
	if err != nil {
		return err
	}
	ApplyMessageSendResult(&state.Leads[index], result, outPath, options.RunID)
	appendRunEvent(&state, RunEvent{
		At:      time.Now(),
		RunID:   options.RunID,
		Phase:   "send-message",
		Bucket:  bucketForLead(state.Leads[index]),
		LeadID:  state.Leads[index].ID,
		Name:    state.Leads[index].Name,
		Result:  result.Status,
		Note:    diagnosticSummary(result),
		OutPath: outPath,
	})
	if err := store.Save(state); err != nil {
		return err
	}
	fmt.Printf("lead=%s status=%s dry_run=%t out=%s\n", lead.ID, result.Status, result.DryRun, outPath)
	return nil
}

func messageSubject(lead Lead) string {
	switch lead.LeadType {
	case LeadTypeAgencyResource, LeadTypeAgencyDelivery, LeadTypeAgencyFounder:
		return "Full-Stack Product Engineer Available for Project Work"
	default:
		return "Full-Stack + AI Product Engineer | Open to Contract Work"
	}
}

func draftSubject(lead Lead) string {
	if lead.Draft != nil && cleanText(lead.Draft.Subject) != "" {
		return strings.TrimSpace(lead.Draft.Subject)
	}
	return messageSubject(lead)
}

func LoadMessageSendResult(path string) (MessageSendResult, error) {
	var result MessageSendResult
	raw, err := os.ReadFile(path)
	if err != nil {
		return MessageSendResult{}, fmt.Errorf("reading message result %s: %w", path, err)
	}
	if err := json.Unmarshal(raw, &result); err != nil {
		return MessageSendResult{}, fmt.Errorf("parsing message result %s: %w", path, err)
	}
	return result, nil
}

func ApplyMessageSendResult(lead *Lead, result MessageSendResult, outPath string, runID string) {
	now := time.Now()
	note := resultNote(result)
	lead.SendAttempts = append(lead.SendAttempts, SendAttempt{
		At:          now,
		RunID:       runID,
		DryRun:      result.DryRun,
		Status:      result.Status,
		ResultURL:   result.URL,
		Note:        note,
		OutPath:     outPath,
		Diagnostics: sendDiagnostics(result),
	})
	lead.MessageStatus = messageStatusForResult(result)
	lead.MessageStatusAt = &now
	lead.UpdatedAt = now
}

func sendDiagnostics(result MessageSendResult) map[string]string {
	diagnostics := map[string]string{}
	if result.ComposerSelector != nil && cleanText(*result.ComposerSelector) != "" {
		diagnostics["composer"] = cleanText(*result.ComposerSelector)
	}
	if len(result.SubjectFill) > 0 && string(result.SubjectFill) != "null" {
		diagnostics["subject"] = compactJSON(result.SubjectFill)
	}
	if len(result.BodyFill) > 0 && string(result.BodyFill) != "null" {
		diagnostics["body"] = compactJSON(result.BodyFill)
	}
	if len(result.Send) > 0 && string(result.Send) != "null" {
		diagnostics["send"] = compactJSON(result.Send)
	}
	if len(result.SendButtons) > 0 && string(result.SendButtons) != "null" {
		diagnostics["send_buttons"] = compactJSON(result.SendButtons)
	}
	if len(result.ConversationCheck) > 0 && string(result.ConversationCheck) != "null" {
		diagnostics["conversation"] = compactJSON(result.ConversationCheck)
	}
	if len(result.Action) > 0 && string(result.Action) != "null" {
		diagnostics["action"] = compactJSON(result.Action)
	}
	return diagnostics
}

func compactJSON(raw json.RawMessage) string {
	if len(raw) == 0 {
		return ""
	}
	var value any
	if err := json.Unmarshal(raw, &value); err != nil {
		return truncateEvidence(string(raw))
	}
	encoded, err := json.Marshal(value)
	if err != nil {
		return truncateEvidence(string(raw))
	}
	return truncateEvidence(string(encoded))
}

func diagnosticSummary(result MessageSendResult) string {
	parts := []string{}
	if result.Reason != nil && cleanText(*result.Reason) != "" {
		parts = append(parts, cleanText(*result.Reason))
	}
	if result.ComposerSelector != nil && cleanText(*result.ComposerSelector) != "" {
		parts = append(parts, "composer "+cleanText(*result.ComposerSelector))
	}
	if len(result.SubjectFill) > 0 && string(result.SubjectFill) != "null" {
		parts = append(parts, "subject "+compactJSON(result.SubjectFill))
	}
	if len(result.BodyFill) > 0 && string(result.BodyFill) != "null" {
		parts = append(parts, "body "+compactJSON(result.BodyFill))
	}
	if len(result.Send) > 0 && string(result.Send) != "null" {
		parts = append(parts, "send "+compactJSON(result.Send))
	}
	return truncateEvidence(strings.Join(parts, "; "))
}

func appendRunEvent(state *OutreachState, event RunEvent) {
	if event.At.IsZero() {
		event.At = time.Now()
	}
	state.RunEvents = append(state.RunEvents, event)
	const maxRunEvents = 500
	if len(state.RunEvents) > maxRunEvents {
		state.RunEvents = state.RunEvents[len(state.RunEvents)-maxRunEvents:]
	}
}

func messageStatusForResult(result MessageSendResult) MessageStatus {
	switch result.Status {
	case "dry-run-messageable":
		return MessageStatusDryRunReady
	case "sent-clicked":
		return MessageStatusSent
	case "not-messageable":
		return MessageStatusNotMessageable
	case "conversation-exists":
		return MessageStatusConversationExists
	case "blocked":
		return MessageStatusBlocked
	default:
		return MessageStatusSendFailed
	}
}

func resultNote(result MessageSendResult) *string {
	if result.Reason != nil && cleanText(*result.Reason) != "" {
		return result.Reason
	}
	if len(result.Action) > 0 && string(result.Action) != "null" {
		value := string(result.Action)
		return &value
	}
	return nil
}

File: apps/recruiter_agency_outreach/send.py
"""Guarded LinkedIn message send state transitions."""

from __future__ import annotations

import json
from collections.abc import Mapping
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from .dashboard import bucket_for_lead, lead_matches_sendable_bucket
from .drafts import draft_subject
from .models import Lead, MessageStatus, OutreachState, RunEvent, SendAttempt
from .sourcing import find_lead_by_id
from .storage import Store, append_run_event
from .utils import clean_text, now_iso, truncate_evidence


@dataclass(slots=True)
class MessageCandidate:
    id: str
    name: str
    profile_url: str
    source: str
    lead_type: str
    title: str | None = None
    company: str | None = None
    search_url: str | None = None


@dataclass(slots=True)
class SendMessageOptions:
    lead_id: str
    run_id: str = ""
    session: str = ""
    out_dir: str = ""
    dry_run: bool = False
    allow_send: bool = False
    result_path: str = ""
    browser: Any = None


def build_message_candidate(state: OutreachState, lead: Lead) -> MessageCandidate:
    if not lead.profile_url:
        raise ValueError(f"lead {lead.id} has no profile URL")
    search_url = None
    cursor = state.capture_cursors.get(lead.source)
    if cursor and cursor.resume_url:
        search_url = cursor.resume_url
    return MessageCandidate(
        id=lead.id,
        name=lead.name,
        profile_url=lead.profile_url,
        source=lead.source,
        lead_type=lead.lead_type.value,
        title=lead.title,
        company=lead.company,
        search_url=search_url,
    )


def prepare_message_config(
    state: OutreachState, lead: Lead, dry_run: bool, allow_send: bool
) -> dict[str, Any]:
    if lead.draft is None or not clean_text(lead.draft.body):
        raise ValueError(f"lead {lead.id} has no draft; run draft first")
    candidate = build_message_candidate(state, lead)
    return {
        "candidate": {
            "id": candidate.id,
            "name": candidate.name,
            "profileUrl": candidate.profile_url,
            "searchUrl": candidate.search_url,
            "source": candidate.source,
            "leadType": candidate.lead_type,
            "title": candidate.title,
            "company": candidate.company,
        },
        "message": lead.draft.body,
        "subject": draft_subject(lead),
        "dryRun": dry_run,
        "allowSend": allow_send,
    }


def send_message(store: Store, options: SendMessageOptions) -> str:
    run_id = options.run_id or _default_run_id("message")
    state = store.load()
    lead = find_lead_by_id(state.leads, options.lead_id)
    if lead is None:
        raise ValueError(f"unknown lead id {options.lead_id!r}")
    dry_run = options.dry_run or not options.allow_send
    if not options.session:
        raise ValueError("--session is required")
    if lead.draft is None or not clean_text(lead.draft.body):
        raise ValueError(f"lead {lead.id} has no draft; run draft first")
    if not lead.profile_url:
        raise ValueError(f"lead {lead.id} has no profile URL")
    if not dry_run and lead.message_status != MessageStatus.DRY_RUN_READY:
        raise ValueError(
            f"lead {lead.id} is {lead.message_status.value}; real sends require dry_run_ready"
        )
    bucket = bucket_for_lead(lead)
    if not dry_run and not lead_matches_sendable_bucket(state, lead, bucket):
        raise ValueError(
            f"lead {lead.id} is not sendable for {bucket}; "
            "agency sends require qualified account context"
        )
    if not dry_run and not options.allow_send:
        raise ValueError("real send requires --allow-send")

    config = prepare_message_config(state, lead, dry_run, options.allow_send)
    if options.result_path:
        result = load_message_send_result(options.result_path)
        out_path = options.result_path
    else:
        browser = options.browser or _default_message_browser(options, store)
        try:
            try:
                result, out_path = browser.send_message(
                    config,
                    dry_run=dry_run,
                    allow_send=options.allow_send,
                )
            except Exception as exc:
                if dry_run:
                    raise
                out_path = _write_browser_exception_result(
                    browser,
                    store,
                    lead,
                    dry_run=dry_run,
                    exc=exc,
                )
                result = load_message_send_result(out_path)
        finally:
            close = getattr(browser, "close", None)
            if callable(close):
                close()
    if not result.dry_run and dry_run:
        raise ValueError("real send result requires --allow-send")
    if result.status == "sent-clicked" and result.dry_run:
        raise ValueError("sent-clicked result cannot be dry_run=true")
    apply_message_send_result(lead, result, out_path, run_id)
    append_run_event(
        state,
        RunEvent(
            at=now_iso(),
            run_id=run_id,
            phase="send-message",
            command="send-message",
            bucket=bucket,
            lead_id=lead.id,
            name=lead.name,
            result=result.status,
            note=result_note(result) or "",
            out_path=out_path,
            state_path=str(store.state_path),
        ),
    )
    store.save(state)
    return (
        f"lead={lead.id} status={result.status} "
        f"dry_run={str(result.dry_run).lower()} out={out_path}"
    )


def _default_message_browser(options: SendMessageOptions, store: Store) -> Any:
    from .message_browser import PlaywriterMessageBrowserClient

    out_dir = Path(options.out_dir) if options.out_dir else store.dir / "message-results"
    session = None if options.session == "auto" else options.session
    return PlaywriterMessageBrowserClient(out_dir=out_dir, session=session)


def _write_browser_exception_result(
    browser: Any,
    store: Store,
    lead: Lead,
    *,
    dry_run: bool,
    exc: Exception,
) -> str:
    out_dir = Path(getattr(browser, "out_dir", store.dir / "message-results"))
    out_dir.mkdir(parents=True, exist_ok=True)
    path = out_dir / f"{now_iso().replace(':', '').replace('-', '')}-{_safe_stem(lead.id)}.json"
    payload = {
        "status": "send-failed",
        "dryRun": dry_run,
        "url": lead.profile_url,
        "reason": f"{type(exc).__name__}: {exc}",
    }
    path.write_text(json.dumps(payload, indent=2, sort_keys=True) + "\n")
    return str(path)


def _safe_stem(value: str) -> str:
    cleaned = []
    for char in value:
        cleaned.append(char if char.isalnum() or char in {"-", "_"} else "_")
    return "".join(cleaned).strip("._") or "lead"


@dataclass(slots=True)
class MessageSendResult:
    status: str
    dry_run: bool = True
    url: str | None = None
    reason: str | None = None
    body: str | None = None
    composer_selector: str | None = None
    action: dict[str, Any] | None = None
    search_row_action: dict[str, Any] | None = None
    conversation_check: dict[str, Any] | None = None
    subject_fill: dict[str, Any] | None = None
    body_fill: dict[str, Any] | None = None
    send: dict[str, Any] | None = None
    send_buttons: list[dict[str, Any]] | None = None

    @classmethod
    def from_mapping(cls, data: Mapping[str, Any]) -> MessageSendResult:
        return cls(
            status=str(data.get("status") or ""),
            dry_run=bool(data.get("dryRun", data.get("dry_run", True))),
            url=_optional_str(data.get("url")),
            reason=_optional_str(data.get("reason")),
            body=_optional_str(data.get("body")),
            composer_selector=_optional_str(data.get("composerSelector")),
            action=_optional_dict(data.get("action")),
            search_row_action=_optional_dict(data.get("searchRowAction")),
            conversation_check=_optional_dict(data.get("conversationCheck")),
            subject_fill=_optional_dict(data.get("subjectFill")),
            body_fill=_optional_dict(data.get("bodyFill")),
            send=_optional_dict(data.get("send")),
            send_buttons=_optional_list_dict(data.get("sendButtons")),
        )


def load_message_send_result(path: str | Path) -> MessageSendResult:
    raw = json.loads(Path(path).read_text())
    if not isinstance(raw, dict):
        raise ValueError(f"{path} did not contain a JSON object")
    return MessageSendResult.from_mapping(raw)


def apply_message_send_result(
    lead: Lead,
    result: MessageSendResult,
    out_path: str,
    run_id: str,
) -> None:
    at = now_iso()
    lead.send_attempts.append(
        SendAttempt(
            at=at,
            run_id=run_id,
            dry_run=result.dry_run,
            status=result.status,
            result_url=result.url,
            note=result_note(result),
            out_path=out_path,
            diagnostics=send_diagnostics(result),
        )
    )
    lead.message_status = message_status_for_result(result)
    lead.message_status_at = at
    lead.updated_at = at


def message_status_for_result(result: MessageSendResult) -> MessageStatus:
    if result.status == "dry-run-messageable":
        return MessageStatus.DRY_RUN_READY
    if result.status == "sent-clicked":
        return MessageStatus.SENT
    if result.status == "not-messageable":
        return MessageStatus.NOT_MESSAGEABLE
    if result.status == "conversation-exists":
        return MessageStatus.CONVERSATION_EXISTS
    if result.status == "blocked":
        return MessageStatus.BLOCKED
    return MessageStatus.SEND_FAILED


def result_note(result: MessageSendResult) -> str | None:
    if result.reason:
        return result.reason
    if result.action:
        return truncate_evidence(json.dumps(result.action, sort_keys=True))
    return None


def send_diagnostics(result: MessageSendResult) -> dict[str, str]:
    diagnostics: dict[str, str] = {}
    if result.composer_selector:
        diagnostics["composer"] = result.composer_selector
    for key, value in {
        "subject": result.subject_fill,
        "body": result.body_fill,
        "send": result.send,
        "conversation": result.conversation_check,
        "action": result.action,
    }.items():
        if value:
            diagnostics[key] = truncate_evidence(json.dumps(value, sort_keys=True))
    if result.send_buttons:
        diagnostics["send_buttons"] = truncate_evidence(
            json.dumps(result.send_buttons, sort_keys=True)
        )
    return diagnostics


def _optional_str(value: object) -> str | None:
    cleaned = clean_text(value)
    return cleaned or None


def _optional_dict(value: object) -> dict[str, Any] | None:
    return dict(value) if isinstance(value, Mapping) else None


def _optional_list_dict(value: object) -> list[dict[str, Any]] | None:
    if not isinstance(value, list):
        return None
    return [dict(item) for item in value if isinstance(item, Mapping)]


def _default_run_id(prefix: str) -> str:
    stamp = now_iso().replace(":", "").replace("-", "").replace(".", "")
    return f"{prefix}-{stamp}"
